include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -2124,6 +2275,180 @@ __parallel_partial_sort(oneapi::dpl::__internal::__device_backend_tag __backend_
+struct __reduce2_wrapper;
+
+template <typename _Name>
+struct __assign_key1_wrapper;
+
+template <typename _Name>
+struct __assign_key2_wrapper;
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__difference_t<_Range3>
+__parallel_reduce_by_segment(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec, _Range1&& __keys,
+                             _Range2&& __values, _Range3&& __out_keys, _Range4&& __out_values,
+                             _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    // The algorithm reduces values in __values where the
+    // associated keys for the values are equal to the adjacent key.
+    //
+    // Example: __keys       = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 }
+    //          __values     = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 }
+    //
+    //          __out_keys   = { 1, 2, 3, 4, 1, 3, 1, 3, 0 }
+    //          __out_values = { 1, 2, 3, 4, 2, 6, 2, 6, 0 }
+
+    const auto __n = __keys.size();
+
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+    // Prior to icpx 2025.0, the reduce-then-scan path performs poorly and should be avoided.
+#if !defined(__INTEL_LLVM_COMPILER) || __INTEL_LLVM_COMPILER >= 20250000
+    if constexpr (std::is_trivially_copyable_v<__val_type>)
+    {
+        if (oneapi::dpl::__par_backend_hetero::__is_gpu_with_sg_32(__exec))
+        {
+            auto __res = oneapi::dpl::__par_backend_hetero::__parallel_reduce_by_segment_reduce_then_scan(
+                oneapi::dpl::__internal::__device_backend_tag{}, std::forward<_ExecutionPolicy>(__exec),
+                std::forward<_Range1>(__keys), std::forward<_Range2>(__values), std::forward<_Range3>(__out_keys),
+                std::forward<_Range4>(__out_values), __binary_pred, __binary_op);
+            __res.wait();
+            // Because our init type ends up being tuple<std::size_t, ValType>, return the first component which is the write index. Add 1 to return the
+            // past-the-end iterator pair of segmented reduction.
+            return std::get<0>(__res.get()) + 1;
+        }
+    }
+#endif
+    if constexpr (oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOperator, __val_type>::value)
+    {
+        return __parallel_reduce_by_segment_known_identity(
+            oneapi::dpl::__internal::__device_backend_tag{}, std::forward<_ExecutionPolicy>(__exec),
+            std::forward<_Range1>(__keys), std::forward<_Range2>(__values), std::forward<_Range3>(__out_keys),
+            std::forward<_Range4>(__out_values), __binary_pred, __binary_op);
+    }
+    else
+    {
+        // Round 1: reduce with extra indices added to avoid long segments
```

Comment:
This is perhaps just a preference thing for code organization, but perhaps we can name this implementation and put it in its own function, that way this function can just be the algorithm selection, and it may be more clear.

Reply:
I think this is a good idea and will make this change.

Reply:
I have made this change. The two fallback paths I have named to `__parallel_reduce_by_segment_fallback` with a true_type / false_type parameter for whether there is a known identity.

I do not like that one fallback path is in `parallel_backend_sycl.h` and the other is in `parallel_backend_sycl_reduce_by_segment.h`. However, I did this since I did not want to add both fallback paths into `parallel_backend_sycl_reduce_by_segment.h` as it would require forward declarations for the other parallel algorithms calls made, and I also did not want to add another 400+ lines to `parallel_backend_sycl.h`